### PR TITLE
Use new Preprocessing

### PR DESCRIPTION
**Pull request by nikicc**
Modifications in accordance with new preprocessing:
* Added `tokens` property to `Corpus`. If some preprocessing was done beforehand use it else run default Preprocessor.
* Adopted LDA & WordCloud to use `Corpus.tokens`
* BoW now makes a copy of Corpus.

### CODE REVIEW

File: orangecontrib/text/corpus.py [hunk 1]
```diff
@@ -88,6 +90,14 @@ def _infer_text_features(self):
         self.set_text_features(include_feats)
 
     def extend_corpus(self, metadata, Y):
```
Comment (kernc) [inline orangecontrib/text/corpus.py]: Can `Y` be `None` by default? This is conceptually kind of like `Table.concatenate(..., axis=1)`?

File: orangecontrib/text/corpus.py [hunk 2]
```diff
@@ -126,6 +138,25 @@ def documents(self):
...
+    @property
+    def tokens(self):
+        """
+        Return a list of lists containing tokens. If tokens are not yet
+        present, run default preprocessor and save tokens.
+        """
+        if self._tokens is None:
+            from orangecontrib.text.preprocess import Preprocessor
+            p = Preprocessor()
+            self._tokens = p(self.documents)
```
Comment (kernc) [inline orangecontrib/text/corpus.py]: After one uses `extend_corpus()` method, additional `metas` become available, `documents` property considers them, but `_tokens` aren't invalidated/updated.

File: orangecontrib/text/lda.py [hunk 3]
```diff
@@ -36,10 +37,11 @@ def __init__(self, text, num_topics=5, callback=None):
 
         self.lda = models.LdaModel(id2word=dictionary, num_topics=self.num_topics)
         done = 0
-        for i, part in enumerate(chunk_list(corpus, 95)):
+        for i, part in enumerate(chunk_list(corpus, 100)):
             self.lda.update(part)
             done += len(part)
-            if callback: callback(95.0*done/len(corpus))
+            if callback:
+                callback(len(part))
```
Comment (kernc) [inline orangecontrib/text/lda.py]: `done` is no longer used. Before, `callback` was called with "done ratio". Now it's with "current batch size"?

File: orangecontrib/text/widgets/owwordcloud.py [hunk 4]
```diff
@@ -215,39 +210,25 @@ def on_topics_change(self, data):
...
-        col = self._get_text_column()
-        if col is None: return
-        words = self.PREPROCESS.cv.get_feature_names()
-        freqs = np.array(self.bow.sum(axis=0))[0]
-        self._repopulate_wordcloud(words, freqs)
+    def _apply_corpus(self):
+        words, freq = [], []
+        counts = self.corpus_counter.most_common()
+        if counts:
+            words, freq = zip(*counts)
```
Comment (kernc) [inline orangecontrib/text/widgets/owwordcloud.py]: Can reduce three lines into:
```py
words, freq = zip(*counts) if counts else [], []
```

File: orangecontrib/text/corpus.py (hunk 1 above)
Comment (nikicc) [inline orangecontrib/text/corpus.py]: Could be. However, this is kind of tailored for NYT where `Y` is always present. To make it more general, we should also consider `X`. I suggest this to be done in another PR.

About `Table.concatenate(..., axis=1)`, maybe `extend` would be more appropriate when it is fixed 😄

File: orangecontrib/text/lda.py (hunk 3 above)
Comment (nikicc) [inline orangecontrib/text/lda.py]: Yup, now it's with "current batch size" since callback is `progressBar.advance`.

**Comment by nikicc** [comment]
I think all is fixed. @kernc could you check again?

**Comment by kernc** [comment]
Is `_tokens` cache reset when corpus is modified?

**Comment by nikicc** [comment]
Not yet 😞 Will look into it.

**Comment by nikicc** [comment]
I took an other look and I think it is OK. When using the provided methods like `extend_corpus` and `set_text_features` tokens are invalidated. The only case where inconsistencies could arise is if the user modifies `corpus.X` or `corpus.metas` directly. Since this is not possible through GUI I think we are fine.

File: orangecontrib/text/widgets/owgeomap.py [hunk 5]
```diff
@@ -146,10 +146,11 @@ def on_map_change(self, map_code=''):
                               Map.EUROPE: (INV_CC_EUROPE, CC_EUROPE)} [map_code]
         # Set country counts in JS
         data = defaultdict(int)
-        for cc in self.cc_counts:
-            key = inv_cc_map.get(cc, cc)
-            if key in cc_map:
-                data[key] += self.cc_counts[cc]
+        if hasattr(self, 'cc_counts'):
+            for cc in self.cc_counts:
```
Comment (kernc) [inline orangecontrib/text/widgets/owgeomap.py]:
```py
for cc in getattr(self, 'cc_counts', ()):
    ...
```
